heartbeat/sender/peerAuthenticationSender_test.go

```diff
@@ -0,0 +1,427 @@
+			GetPeerSignatureCalled: func(key crypto.PrivateKey, pid []byte) ([]byte, error) {
+				return nil, expectedErr
+			},
+		}
+		sender, _ := NewPeerAuthenticationSender(args)
+
+		err := sender.execute()
+		assert.Equal(t, expectedErr, err)
+	})
+	t.Run("marshaller fails fot the second time, should return error", func(t *testing.T) {
+		t.Parallel()
+
+		numCalls := 0
+		args := createMockPeerAuthenticationSenderArgs()
+		args.Messenger = &mock.MessengerStub{
+			BroadcastCalled: func(topic string, buff []byte) {
+				assert.Fail(t, "should have not called Messenger.BroadcastCalled")
+			},
+		}
+		args.Marshaller = &mock.MarshallerStub{
+			MarshalHandler: func(obj interface{}) ([]byte, error) {
+				numCalls++
+				if numCalls < 2 {
+					return make([]byte, 0), nil
+				}
+				return nil, expectedErr
+			},
+		}
+		sender, _ := NewPeerAuthenticationSender(args)
+
+		err := sender.execute()
+		assert.Equal(t, expectedErr, err)
+	})
+	t.Run("should work with stubs", func(t *testing.T) {
+		t.Parallel()
+
+		args := createMockPeerAuthenticationSenderArgs()
+		broadcastCalled := false
+		args.Messenger = &mock.MessengerStub{
+			BroadcastCalled: func(topic string, buff []byte) {
+				assert.Equal(t, args.Topic, topic)
+				broadcastCalled = true
+			},
+		}
+		sender, _ := NewPeerAuthenticationSender(args)
+
+		err := sender.execute()
+		assert.Nil(t, err)
+		assert.True(t, broadcastCalled)
+	})
+	t.Run("should work with some real components", func(t *testing.T) {
+		t.Parallel()
+
+		startTime := time.Now()
+		// use the Elrond defined ed25519 operations instead of the secp256k1 implemented in the "real" network messenger,
+		// should work with both
+		keyGen := signing.NewKeyGenerator(ed25519.NewEd25519())
+		skMessenger, pkMessenger := keyGen.GeneratePair()
+		signerMessenger := ed25519SingleSig.Ed25519Signer{}
+
+		args := createMockPeerAuthenticationSenderArgsSemiIntegrationTests()
+		var buffResulted []byte
+		messenger := &mock.MessengerStub{
+			BroadcastCalled: func(topic string, buff []byte) {
+				assert.Equal(t, args.Topic, topic)
+				buffResulted = buff
+			},
+			SignCalled: func(payload []byte) ([]byte, error) {
+				return signerMessenger.Sign(skMessenger, payload)
+			},
+			VerifyCalled: func(payload []byte, pid core.PeerID, signature []byte) error {
+				pk, _ := keyGen.PublicKeyFromByteArray(pid.Bytes())
+
+				return signerMessenger.Verify(pk, payload, signature)
+			},
+			IDCalled: func() core.PeerID {
+				pkBytes, _ := pkMessenger.ToByteArray()
+				return core.PeerID(pkBytes)
+			},
+		}
+		args.Messenger = messenger
+		sender, _ := NewPeerAuthenticationSender(args)
+
+		err := sender.execute()
+		assert.Nil(t, err)
+
+		skBytes, _ := sender.privKey.ToByteArray()
+		pkBytes, _ := sender.publicKey.ToByteArray()
+		log.Info("args", "pid", args.Messenger.ID().Pretty(), "bls sk", skBytes, "bls pk", pkBytes)
+
+		// verify the received bytes if they can be converted in a valid peer authentication message
```

Comment:
you can transform these comments into t.Run("description")

Reply:
can be done, sure

Reply:
done